geomstats/geometry/trees.py

```diff
@@ -0,0 +1,485 @@
+        """
+        class MyCheckException(Exception):
+            pass
+        xs = [set(x) for x in self.partition]
+        ys = [set(y) for y in other.partition]
+        # ----- check out condition 1. of the partial ordering -----
+        try:
+            x_to_y = {i: [j for j, y in enumerate(ys) if x.issubset(y)][0] for i, x in
+                      enumerate(xs)}
+        except IndexError:  # in this case, x is not a refinement of y.
+            return False
+        # ----- check out condition 2. of the partial ordering -----
+        try:
+            for i, splits in enumerate(self.split_sets):
+                # check if the splits are contained in the other restricted splits of
+                # the corresponding component.
+                restr_other_splits = {sp_y.restr(subset=xs[i]) for sp_y in
+                                      other.split_sets[x_to_y[i]]}
+                if not set(splits).issubset(restr_other_splits):
+                    raise MyCheckException()
+        except MyCheckException:
+            return False
+        # ----- check out condition 3. of the partial ordering -----
+        try:
+            for j in range(len(ys)):
+                xs_in_y = [x for i, x in enumerate(xs) if x_to_y[i] == j]
+                for x1, x2 in it.combinations(xs_in_y, r=2):
+                    sep_sp = [sp for sp in other.split_sets[j] if sp.separates(x1, x2)]
+                    if len(sep_sp) == 0:
+                        raise MyCheckException()
+        except MyCheckException:
+            return False
+
+        # ----- all conditions are satisfied -> it is indeed a partial ordering! -----
+        return True
+
+    def __gt__(self, other):
+        """ Strictly greater than (>). """
+        return other < self
+
+    def __ge__(self, other):
+        """ Greater than or equal function (>=). """
+        return other <= self
+
+    def __lt__(self, other):
+        """ Less than function (<). """
+        return self <= other and self != other
+
+    def __hash__(self):
+        """ Computes the hash of the structure. """
+        return hash(str(self))
+
+
+class Wald:
+    r""" A class for phylogenetic forests, elements of the Wald space.
+
+    Parameters
+    ----------
+    n : int
+        Number of labels, the set of labels is then :math:`\{0,\dots,n-1\}`.
+    st : Structure
+        The structure of the forest.
+    x : array-like
+        The edge weights, array of floats between 0 and 1, with m entries, where m is
+        the total number of splits/edges in the structure ``st``.
+    """
+
+    def __init__(self, n, st: Structure, x):
+        self._n: int = n
+        self._st: Structure = st
+        self._x = x
+        self._corr = None
+
+    @property
+    def st(self) -> Structure:
+        """ The structure of the wald. """
+        return self._st
+
+    @property
+    def x(self) -> gs.array:
+        """ The (flat) vector containing the edge weights in Nye notation. """
```

Comment:
Nye notation should be maybe substituted with a reference, otherwise if it is an internal lexicon we should better explain it